include/cuda/memory_resource

```diff
@@ -0,0 +1,549 @@
+class basic_resource_ref;
+
+template <class... _Properties>
+struct _Resource_vtable : public _Property_vtable<_Properties>...
+{
+  _Resource_vtable() = default;
+
+  template <class... _PropertyFns>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr _Resource_vtable(_PropertyFns... __property_fn_) noexcept
+      : _Property_vtable<_Properties>(__property_fn_)...
+  {}
+  
+  template <_AllocType _Alloc_type, class... _OtherProperties>
+  _LIBCUDACXX_INLINE_VISIBILITY constexpr _Resource_vtable(basic_resource_ref<_Alloc_type, _OtherProperties...>& __ref) noexcept
+      : _Property_vtable<_Properties>(__ref._Property_vtable<_Properties>::__property_fn)...
+  {}
+
+  template <class _Resource>
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr _Resource_vtable _Create() noexcept
+  {
+    return {&_Resource_vtable_builder::_Get_property<_Resource, _Properties>...};
+  }
+};
+
+template <class... _Properties>
+struct _Filtered;
+
+template <bool _IsUniqueProperty>
+struct _Property_filter
+{
+  template <class _Property, class... _Properties>
+  using _Filtered_properties =
+    typename _Filtered<_Properties...>::_Filtered_vtable::template _Append_property<_Property>;
+};
+template <>
+struct _Property_filter<false>
+{
+  template <class _Property, class... _Properties>
+  using _Filtered_properties = typename _Filtered<_Properties...>::_Filtered_vtable;
+};
+
+template <class _Property, class... _Properties>
+struct _Filtered<_Property, _Properties...>
+{
+  using _Filtered_vtable =
+    typename _Property_filter<property_with_value<_Property> && !_CUDA_VSTD::_One_of<_Property, _Properties...>>::
+      template _Filtered_properties<_Property, _Properties...>;
+
+  template <class _OtherPropery>
+  using _Append_property = _Filtered<_OtherPropery, _Property, _Properties...>;
+
+  using _Vtable = _Resource_vtable<_Property, _Properties...>;
+};
+
+template <>
+struct _Filtered<>
+{
+  using _Filtered_vtable = _Filtered<>;
+
+  template <class _OtherPropery>
+  using _Append_property = _Filtered<_OtherPropery>;
+
+  using _Vtable = _Resource_vtable<>;
+};
+
+template <class... _Properties>
+using _Filtered_vtable = typename _Filtered<_Properties...>::_Filtered_vtable::_Vtable;
+
+struct _Alloc_base {
+public:
+    _LIBCUDACXX_INLINE_VISIBILITY _Alloc_base(void* __object_, const _Async_alloc_vtable* __static_vtabl_) noexcept
+      : __object(__object_), __static_vtable(__static_vtabl_) {}
+
+    _LIBCUDACXX_INLINE_VISIBILITY void* allocate(size_t __bytes, size_t __alignment) {
+        return __static_vtable->__alloc_fn(__object, __bytes, __alignment);
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY void deallocate(void* _Ptr, size_t __bytes, size_t __alignment) {
+        __static_vtable->__dealloc_fn(__object, _Ptr, __bytes, __alignment);
+    }
+
+protected:
+  void* __object                             = nullptr;
+  const _Async_alloc_vtable* __static_vtable = nullptr;
+};
+
+struct _Async_alloc_base : public _Alloc_base { 
```

Comment:
```suggestion
template <typename _Vtable = _Async_alloc_vtable>
struct _Async_alloc_base : public _Alloc_base<_Vtable> { 
```

Reply:
Ditto, this change adds complexity that does not bring any benefits